site/app/controllers/student/SubmissionController.php

```diff
@@ -1020,6 +1020,75 @@ public function ajaxRegrade($gradeable_id): array {
         return $this->core->getOutput()->renderResultMessage($msg, true);
     }
 
+    /**
+     * @return JsonResponse
+     * @param string $gradeable_id
+     */
+    #[Route('/api/{_semester}/{_course}/gradeable/{gradeable_id}/values', methods: ['GET'])]
+    public function ajaxGetGradeableValues(string $gradeable_id): JsonResponse {
+        if ($this->core->getQueries()->getUserById($_GET['user_id']) === null) {
+            return JsonResponse::getFailResponse('User with id `' . $_GET['user_id'] . '` does not exist.');
+        }
+        // Faculty = 2, Superuser = 1
+        if ($this->core->getUser()->getAccessLevel() > 2 && ($_GET['user_id'] ?? '') !== $this->core->getUser()->getId()) {
+            return JsonResponse::getFailResponse('API key and specified user_id are not for the same user.');
+        }
+
+        try {
+            $gradeable = $this->core->getQueries()->getGradeableConfig($gradeable_id);
+        }
+        catch (\InvalidArgumentException $e) {
+            return JsonResponse::getFailResponse('Gradeable does not exist');
+        }
+
+        $graded_gradeable = $this->core->getQueries()->getGradedGradeable(
+            $gradeable,
+            $_GET['user_id'],
+            $gradeable->isTeamAssignment()
+        );
+
+        if ($graded_gradeable === null) {
+            return JsonResponse::getFailResponse('Gradeable does not exist');
+        }
+
+        $graded_gradeable = $graded_gradeable->getAutoGradedGradeable();
+
+        return JsonResponse::getSuccessResponse([
+            'is_queued' => $graded_gradeable->isQueued(),
+            'queue_position' => $graded_gradeable->getQueuePosition(),
+            'is_grading' => $graded_gradeable->isGrading(),
+            'has_submission' => $graded_gradeable->hasSubmission(),
+            'autograding_complete' => $graded_gradeable->isAutoGradingComplete(),
+            'has_active_version' => $graded_gradeable->hasActiveVersion(),
+            'highest_version' => $graded_gradeable->getHighestVersion(),
+            'total_points' => ($graded_gradeable->hasActiveVersion() ? $graded_gradeable->getTotalPoints() : null),
+            'total_percent' => ($graded_gradeable->hasActiveVersion() ? $graded_gradeable->getTotalPercent() : null)
+        ]);
+    }
+
+    /**
+     * @return JsonResponse|array{
+     *     status: string,
+     *     data: mixed
+     * }
+     */
+    #[Route('/api/{_semester}/{_course}/gradeable/{gradeable_id}/grade', methods: ['POST'])]
+    public function ajaxRequestGrade(string $gradeable_id): JsonResponse|array {
+        // Faculty = 2, Superuser = 1
+        if ($this->core->getUser()->getAccessLevel() > 2 && ($_POST['user_id'] ?? '') !== $this->core->getUser()->getId()) {
```

Comment:
Instead of magic numbers, it would be better to reference static constants if possible.